src/snowflake/connector/cache.py

```diff
@@ -0,0 +1,264 @@
+        self,
+        key: K,
+    ) -> None:
+        """Non-locking version of __delitem__.
+
+        This should only be used by internal functions when already
+        holding self._lock.
+        """
+        del self._cache[key]
+        self.telemetry["size"] = len(self._cache)
+
+    def __delitem__(
+        self,
+        key: K,
+    ) -> None:
+        with self._lock:
+            self.__delitem(key)
+
+    def __contains__(
+        self,
+        key: K,
+    ) -> bool:
+        with self._lock:
+            if key in self._cache.keys():
+                try:
+                    self.__getitem(key, should_record_hits=True)
+                    return True
+                except KeyError:
+                    # Fall through
+                    pass
+        return False
+
+    def update(
+        self,
+        other: dict[K, V] | SFDictCache[K, V],
+    ) -> None:
+        """Insert multiple values at the same time.
+
+        If this function is given a dictionary, expiration timestamps
+        will be all the same a self._entry_lifetime form now. If it's
+        given another SFDictCache then the timestamps will be taken
+        from the other cache.
+        """
+        if isinstance(other, dict):
+            t = now() + self._entry_lifetime
+            to_insert: dict[K, CacheEntry[V]] = {
+                k: CacheEntry(expiry=t, entry=v) for k, v in other.items()
+            }
+            with self._lock:
+                self._cache.update(to_insert)
+                self.telemetry["size"] = len(self._cache)
+        elif isinstance(other, SFDictCache):
+            to_insert: dict[K, CacheEntry[V]] = {k: v for k, v in other._cache.items()}
+            with self._lock:
+                self._cache.update(to_insert)
+                self.telemetry["size"] = len(self._cache)
+        else:
+            raise TypeError
+
+    def _clear_expired_entries(self) -> None:
+        with self._lock:
+            for k in self._cache.keys():
+                try:
+                    self.__getitem(k, should_record_hits=False)
+                except KeyError:
+                    continue
+            self.telemetry["size"] = len(self._cache)
+
+    # Telemetry related functions, these can be plugged by child classes
+    def _reset_telemetry(self) -> None:
+        """(Re)set telemetry fields.
+
+        This function will be called by the initalizer and other functions that should
+        reset telemtry entries.
+        """
+        self.telemetry = {
+            "hit": 0,
+            "miss": 0,
+            "expiration": 0,
+            "size": 0,
+        }
+
+    def _hit(self, k: K) -> None:
+        """This function gets called when a hit occurs.
+
+        Functions that hit every entry (like values) is not going to count.
+
+        Note that while this function does not interact with lock, but it's only
+        called from contexts where the lock is already held.
+        """
+        del k
```

Comment:
may I ask why `del K` is called here?

Reply:
To get around linters complaining about this argument being unused

Reply:
I see, I think there should be a better way to ignore the lint error, maybe try adding `# noqa` on the line where the error is reported? https://flake8.pycqa.org/en/3.1.1/user/ignoring-errors.html#in-line-ignoring-errors

I'm not sure what side effect that del K could give us (e.g. reduce ref cnt xx)

Reply:
It does reduce ref count, but it should only delete the local scope's ref and not the outer scope, but I hear you.
I'll change it for something else!